### ISSUE TITLE
BLD: can we remove the upper bound on Cython (`<3.1`) ?

### ISSUE THREAD

**Issue by neutrinoceros** [astropy/astropy#16751]
In #15402, @saimn added build-time compat for Cython 3 and set an upper bound to `<3.1`.
In other projects, I've seen this upper bound used as a security measure against some deprecations that were originally intended to end in Cython 3.1 but didn't have a clear migration path (see https://github.com/cython/cython/issues/4310), but in astropy it's not clear to me what's the actual reason for it (`IF` and `DEF` are not used anywhere). I suspect we'll actually *need* Cython 3.1 to start experimenting with the upcoming (experimental) free-threaded version of CPython, so I'm wondering if it's actually safe to remove the pin now. Maybe the safest thing to do is to actually wait for Cython 3.1 to be released and try to remove the pin then ?
For completion I don't have any reason to think that we'll need Cython 3.1 to build against Python 3.13 (GIL version), which is why I'm opening a new ticket separate from #16595

**Comment by neutrinoceros** [comment]
FWIW I just succeeded in building astropy with the latest Cython 3.1 nightly so I'm going to open a PR to propose this change now.

**Comment by saimn** [comment]
The risk is that released versions without a pin might break with some future version of cython. For `main` it's fine to remove the pin but I think we would still need it when releasing. Not sure how other projects handle that.
cc @astrofrog

**Comment by astrofrog** [comment]
In principle I think all build-time dependencies should probably have upper pins, at least to prevent the next major version from getting installed. Take for example the astropy 6.1.2 release I did this week - if someone tries to installing it 5 years, and Cython 4 and Numpy 3 are out and break compatibility, it will be hard to install astropy 6.1.2 from source without disabling the isolated build environment completely.

For Cython, the pin was originally added because in fact we used to ship the generated C code, so pinning was a way of effectively getting the same result, that the generated C code would be deterministic.

Perhaps we should pin ``cython<4``  and ``numpy<3``?
